File: apps/weather/get_data.py
import os
import sys
import requests

from apps.weather.service.data import write_data


"""
use qweather.com service
"""


def call_15d_weather_api(token: str, location_id: str, lang:str='en'):
    """
    location id: https://github.com/qwd/LocationList?tab=readme-ov-file
    """
    res = requests.get(f"https://devapi.qweather.com/v7/weather/15d?location={location_id}&key={token}&lang={lang}")
    write_data(text = res.text, file_name="15d_data")

def call_aqi_api(token: str, location_id: str, lang:str='en'):
    """
    location id: https://github.com/qwd/LocationList?tab=readme-ov-file
    """
    res = requests.get(f"https://devapi.qweather.com/v7/air/5d?location={location_id}&key={token}&lang={lang}")
    write_data(text = res.text, file_name="aqi_data")
    

if __name__ == "__main__":
    try:
        sys.argv[1]
        sys.argv[2]
    except IndexError as e:
        print("use qweather.com service, token and location id is required")
        exit()
    call_15d_weather_api(sys.argv[1], sys.argv[2])
    call_aqi_api(sys.argv[1], sys.argv[2])

File: apps/weather/get_data.go
package main

import (
	"context"
	"fmt"
	"os"
	"sync"
	"weather/service"

	"github.com/carlmjohnson/requests"
)

func Call15DaysWeatherApi(ctx context.Context, token string, location_id string, lang string, wg *sync.WaitGroup) {
	// location id: https://github.com/qwd/LocationList?tab=readme-ov-file
	defer wg.Done()
	var s string
	var err error
	req := requests.URL(fmt.Sprintf("https://devapi.qweather.com/v7/weather/15d?location=%s&key=%s&lang=%s", location_id, token, lang))
	err = req.ToString(&s).Fetch(ctx)
	if err != nil {
		fmt.Printf("Call15DaysWeatherApi failed: %s", err.Error())
		return
	}
	err = service.WriteData(s, "15d_data")
	if err != nil {
		fmt.Printf("write data failed: %s", err.Error())
		return
	}

}

func CallAqiApi(ctx context.Context, token string, location_id string, lang string, wg *sync.WaitGroup) {
	// location id: https://github.com/qwd/LocationList?tab=readme-ov-file
	defer wg.Done()
	var s string
	var err error
	req := requests.URL(fmt.Sprintf("https://devapi.qweather.com/v7/air/5d?location=%s&key=%s&lang=%s", location_id, token, lang))
	err = req.ToString(&s).Fetch(ctx)
	if err != nil {
		fmt.Printf("CallAqiApi failed: %s", err.Error())
		return
	}
	err = service.WriteData(s, "aqi_data")
	if err != nil {
		fmt.Printf("write data failed: %s", err.Error())
		return
	}

}

func main() {
	var wg sync.WaitGroup

	ctx := context.TODO()
	wg.Add(1)
	go Call15DaysWeatherApi(ctx, os.Args[1], os.Args[2], "en", &wg)
	wg.Add(1)
	go CallAqiApi(ctx, os.Args[1], os.Args[2], "en", &wg)

	wg.Wait()
}
